flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2008,23 +2065,173 @@ class ArrayExprLowering {
+        switch (Fortran::lower::lowerIntrinsicArgumentAs(getLoc(), *argLowering,
+                                                         dummy.name)) {
+        case Fortran::lower::LowerIntrinsicArgAs::Value: {
+          PushSemantics(ConstituentSemantics::RefTransparent);
+          auto lambda = genarr(*expr);
+          operands.emplace_back([=](IterSpace iters) { return lambda(iters); });
+        } break;
+        case Fortran::lower::LowerIntrinsicArgAs::Addr: {
+          // Note: assume does not have Fortran VALUE attribute semantics.
+          PushSemantics(ConstituentSemantics::RefOpaque);
+          auto lambda = genarr(*expr);
+          operands.emplace_back([=](IterSpace iters) { return lambda(iters); });
+        } break;
+        case Fortran::lower::LowerIntrinsicArgAs::Inquired:
+          TODO(loc, "intrinsic function with inquired argument");
+          break;
+        }
+      }
+    }
+
+    // Let the intrinsic library lower the intrinsic procedure call
+    return [=](IterSpace iters) {
+      llvm::SmallVector<fir::ExtendedValue> args;
+      for (const auto &cc : operands)
+        args.push_back(cc(iters));
+      return Fortran::lower::genIntrinsicCall(builder, loc, name, retTy, args,
+                                              stmtCtx);
+    };
+  }
+
+  // A procedure reference to a user-defined elemental procedure.
+  CC genUDProcRef(const Fortran::evaluate::ProcedureRef &procRef,
+                  llvm::Optional<mlir::Type> retTy) {
+    using PassBy = Fortran::lower::CallerInterface::PassEntityBy;
+
+    Fortran::lower::CallerInterface caller(procRef, converter);
+    llvm::SmallVector<CC> operands(caller.getNumFIRArguments());
+    auto loc = getLoc();
+    auto callSiteType = caller.genFunctionType();
+    for (const auto &arg : caller.getPassedArguments()) {
+      const auto *actual = arg.entity;
+      auto argTy = callSiteType.getInput(arg.firArgument);
+      if (!actual) {
+        // Optional dummy argument for which there is no actual argument.
+        auto absent = builder.create<fir::AbsentOp>(loc, argTy);
+        operands[arg.firArgument] = [=](IterSpace) { return absent; };
+        continue;
+      }
+      const auto *expr = actual->UnwrapExpr();
+      if (!expr)
+        TODO(loc, "assumed type actual argument lowering");
+
+      LLVM_DEBUG(expr->AsFortran(llvm::dbgs()
+                                 << "argument: " << arg.firArgument << " = [")
+                 << "]\n");
+      switch (arg.passBy) {
+      case PassBy::Value: {
+        // True pass-by-value semantics.
+        PushSemantics(ConstituentSemantics::RefTransparent);
+        auto lambda = genarr(*expr);
+        operands[arg.firArgument] = [=](IterSpace iters) {
+          return lambda(iters);
+        };
+      } break;
+      case PassBy::ValueAttribute: {
+        // VALUE attribute or pass-by-reference to a copy semantics. (byval*)
+        PushSemantics(ConstituentSemantics::ByValueArg);
```

Comment:
What is the benefit of dealing with the value case in `genarr(const Fortran::semantics::SymbolRef &sym)` rather than to use the `RefTransparent` semantics and  create the temp here ?

Reply:
I wanted to keep the different semantics flavors localized to one place.